File: 0067-add-binary/0067-add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        char[] aa = a.toCharArray();
        char[] ba = b.toCharArray();
        int r = 0;
        int i = a.length() - 1, j = b.length() - 1;
        while (i >= 0 || j >= 0) {
            int d1 = i >= 0 ? aa[i] - '0' : 0;
            int d2 = j >= 0 ? ba[j] - '0' : 0;
            int d = d1 + d2 + r;
            sb.append(d % 2);
            r = d / 2;
            i--;
            j--;
        }
        if (r == 1) {
            sb.append('1');
        }
        return sb.reverse().toString();
    }
}

File: 0067-add-binary/0067-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i, j, m, n = 0, 0, len(a), len(b)
        carry = 0
        res = []
        while i < m or j < n:
            d1 = 0 if i >= m else int(a[m - i - 1])
            d2 = 0 if j >= n else int(b[n - j - 1])
            d = d1 + d2 + carry
            res.append(d % 2)
            carry = 1 if d > 1 else 0
            i += 1
            j += 1
        if carry == 1:
            res.append(1)
        return ''.join(map(str, res[::-1]))